File: python/largest-series-product/largest_series_product.py
from operator import mul

def largest_product(series, span):
	if span > len(series) or span < 0:
		raise ValueError("Invalid span")
	if not span:
		return 1

	ints = [int(c) for c in series]
	r = len(ints) - span + 1
	products = [ints[idx:span+idx] for idx in range(r)]

	return max(map(_prod, products))

def _prod(x):
	return reduce(mul, x, 1)

File: javascript/largest-series-product/largest-series-product.js
const product = (a, b) => a * b;

const isValid = (digits, size) => size >= 0 && digits.every(d => !Number.isNaN(d));

const slices = (digits, size) => {
    return [...Array(digits.length - size + 1).keys()]
        .map(idx => digits.slice(idx, idx + size));
}

class Series {

    constructor(digits) {
        this.digits = [...digits].map(Number);
    }

    largestProduct(size) {
        if (!isValid(this.digits, size)) {
            throw new Error('Invalid input.');
        }
        if (size > this.digits.length) {
            throw new Error('Slice size is too big.');
        }
        return slices(this.digits, size)
            .reduce((acc, slice) => {
                return Math.max(acc, slice.reduce(product, 1));
            }, 0);
    }
}

module.exports = Series;